File: server/msgd/core/payloads/message.py
"""``message.*`` payload schemas (TDD §2.2)."""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, ConfigDict, field_validator

from msgd.core import ids

__all__ = ["MessageCreatedV1"]


class MessageCreatedV1(BaseModel):
    """Payload for ``message.created`` v1 (§2.2).

    ``model_config`` uses ``extra="allow"`` so additive-only v1 changes (§2.3.2)
    round-trip without loss when validated by an older reader.

    Id fields are **format-validated only** — the prefix and ULID validity are
    checked here to catch malformed references early.  Referential *existence*
    (does the message/user/file exist?) is a server-side concern (§3.2) and out
    of scope for this model.
    """

    model_config = ConfigDict(extra="allow")

    message_id: str
    text: str
    #: New format values arrive via a ``type_version`` bump (§2.3), so this
    #: Literal does not violate D9's additive-only rule.
    format: Literal["markdown", "plain"] = "markdown"
    #: First reply *is* the thread (D7); when set, an ``m_`` id in the same stream.
    thread_root_id: str | None = None
    file_ids: list[str] = []
    mentions: list[str] = []

    @field_validator("message_id")
    @classmethod
    def _check_message_id(cls, value: str) -> str:
        if not ids.is_valid_typed_id(value, ids.IdKind.MESSAGE):
            raise ValueError(f"message_id is not a valid m_ id: {value!r}")
        return value

    @field_validator("thread_root_id")
    @classmethod
    def _check_thread_root_id(cls, value: str | None) -> str | None:
        if value is not None and not ids.is_valid_typed_id(value, ids.IdKind.MESSAGE):
            raise ValueError(f"thread_root_id is not a valid m_ id: {value!r}")
        return value

    @field_validator("file_ids")
    @classmethod
    def _check_file_ids(cls, value: list[str]) -> list[str]:
        for fid in value:
            if not ids.is_valid_typed_id(fid, ids.IdKind.FILE):
                raise ValueError(f"file_ids contains an invalid f_ id: {fid!r}")
        return value

    @field_validator("mentions")
    @classmethod
    def _check_mentions(cls, value: list[str]) -> list[str]:
        for uid in value:
            if not ids.is_valid_typed_id(uid, ids.IdKind.USER):
                raise ValueError(f"mentions contains an invalid u_ id: {uid!r}")
        return value

File: web/src/core/payloads/message.ts
/**
 * `message.*` payload schemas — the browser port of
 * `server/msgd/core/payloads/message.py`.
 *
 * Id fields are *format-validated only*: prefix + ULID validity are checked to
 * catch malformed references early. Referential existence (does the
 * message/user/file exist?) is a server-side concern (§3.2), out of scope here.
 */

import { IdKind, isValidTypedId, newMessageId } from '../ids'

/** Payload for `message.created` v1 (§2.2). */
export type MessageCreatedV1 = {
  message_id: string
  text: string
  format: 'markdown' | 'plain'
  thread_root_id: string | null
  file_ids: string[]
  mentions: string[]
}

/** Options for {@link buildMessageCreatedPayload}; defaults mirror the Python model. */
export interface BuildMessageCreatedPayloadOptions {
  text: string
  format?: 'markdown' | 'plain'
  thread_root_id?: string | null
  file_ids?: string[]
  mentions?: string[]
  message_id?: string
}

/**
 * Mint (when absent) and format-validate a `message.created` v1 payload.
 *
 * Mirrors `MessageCreatedV1`'s field validators: `message_id` and any
 * `thread_root_id` are `m_` ids, `file_ids` are `f_` ids, `mentions` are `u_`
 * ids. Defaults: `format` `"markdown"`, `thread_root_id` `null`, `file_ids` `[]`,
 * `mentions` `[]`.
 *
 * @throws {Error} on a malformed id.
 */
export function buildMessageCreatedPayload(
  options: BuildMessageCreatedPayloadOptions,
): MessageCreatedV1 {
  const messageId = options.message_id ?? newMessageId()
  if (!isValidTypedId(messageId, IdKind.MESSAGE)) {
    throw new Error(`message_id is not a valid m_ id: ${messageId}`)
  }

  const threadRootId = options.thread_root_id ?? null
  if (threadRootId !== null && !isValidTypedId(threadRootId, IdKind.MESSAGE)) {
    throw new Error(`thread_root_id is not a valid m_ id: ${threadRootId}`)
  }

  const fileIds = options.file_ids ?? []
  for (const fileId of fileIds) {
    if (!isValidTypedId(fileId, IdKind.FILE)) {
      throw new Error(`file_ids contains an invalid f_ id: ${fileId}`)
    }
  }

  const mentions = options.mentions ?? []
  for (const userId of mentions) {
    if (!isValidTypedId(userId, IdKind.USER)) {
      throw new Error(`mentions contains an invalid u_ id: ${userId}`)
    }
  }

  return {
    message_id: messageId,
    text: options.text,
    format: options.format ?? 'markdown',
    thread_root_id: threadRootId,
    file_ids: fileIds,
    mentions: mentions,
  }
}
